### PR TITLE
core: move census registration into internal

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
move census registration into internal, and preserve tracing bit when rebuilding a MethodDescriptor

### CODE REVIEW

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 1]
```diff
@@ -563,23 +562,78 @@ private Builder() {}
...
+      void onRegister(MethodDescriptor<?, ?> md);
+    }
+
+    /**
+     * Sets a callback for method descriptor builds.  Called for descriptors with
+     * {@link MethodDescriptor#registerForTracing} set.
+     *
+     * @param registerCallback the callback to handle descriptor registration.
+     */
+    static synchronized void setRegisterCallback(RegisterForTracingCallback registerCallback) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Rename it to setRegisterForTracingCallback()

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 2]
```diff
@@ -48,11 +51,13 @@
   private final @Nullable Object schemaDescriptor;
   private final boolean idempotent;
   private final boolean safe;
+  private final boolean registerForTracing;
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/MethodDescriptor.java]: Done

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 3]
```diff
@@ -563,23 +562,78 @@ private Builder() {}
...
+
+    private static volatile RegisterForTracingCallback registerCallback;
+
+    interface RegisterForTracingCallback {
+      void onRegister(MethodDescriptor<?, ?> md);
+    }
+
+    /**
+     * Sets a callback for method descriptor builds.  Called for descriptors with
+     * {@link MethodDescriptor#registerForTracing} set.
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/MethodDescriptor.java]: Done.

File: core/src/main/java/io/grpc/MethodDescriptor.java (hunk 1 above)
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/MethodDescriptor.java]: Done.

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 4]
```diff
@@ -563,23 +562,78 @@ private Builder() {}
...
+      checkState(Registrations.registerCallback == null, "callback already present");
+      Registrations.registerCallback = checkNotNull(registerCallback, "registerCallback");
+      for (WeakReference<MethodDescriptor<?, ?>> mdRef : pendingRegistrations) {
+        MethodDescriptor<?, ?> md = mdRef.get();
+        if (md != null) {
+          mdRef.clear();
+          registerCallback.onRegister(md);
+        }
+      }
+      drainUnregisteredQueue();
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/MethodDescriptor.java]: Done.

**Pull request merged**